File: FleaMarketGuide/FleaMarketGuide/Classes/NewFeature/Controller/NewFeatureViewController.m
//
//  NewFeatureViewController.m
//  FleaMarketGuide
//
//  Created by Alin on 16/6/12.
//  Copyright © 2016年 Alin. All rights reserved.
//

#import "NewFeatureViewController.h"
#import "NewFeatureCell.h"
#import "ALinFlowLayout.h"
#import "ViewController.h"


@interface NewFeatureViewController ()
/** pageControl */
@property (nonatomic, strong) UIPageControl *pageControl;
/** 当前视频是否播放完成 */
@property (nonatomic, assign, getter=isMovieFinished) BOOL movieFinished;
@end

@implementation NewFeatureViewController

static NSString * const reuseIdentifier = @"NewFeatureCell";

#pragma mark - lazy loading

- (UIPageControl *)pageControl
{
    if (!_pageControl) {
        CGFloat width = 120;
        CGFloat height = 30;
        CGFloat x = (MS_Width - width) * 0.5;
        CGFloat y = MS_Height - 30 - 20;
        UIPageControl *pageControl = [[UIPageControl alloc] initWithFrame:CGRectMake(x, y, width, height)];
        pageControl.pageIndicatorTintColor = [UIColor lightGrayColor];
        pageControl.currentPageIndicatorTintColor = [UIColor redColor];
        [Key_Window addSubview:pageControl];
        
        _pageControl = pageControl;
    }
    return _pageControl;
}

#pragma mark - life circle 

- (instancetype)init
{
    return [super initWithCollectionViewLayout:[[ALinFlowLayout alloc] init]];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setup];
}

- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    self.pageControl.numberOfPages = self.guideImagesArr.count;
}

- (void)dealloc
{
    [self.pageControl removeFromSuperview];
    self.pageControl = nil;
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark - private method

- (void)setup
{
    [self.collectionView registerClass:[NewFeatureCell class] forCellWithReuseIdentifier:reuseIdentifier];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(movieFinished) name:PlayFinishedNotify object:nil];
}

- (void)movieFinished
{
    self.movieFinished = YES;
}


#pragma mark <UICollectionViewDataSource>

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {

    return self.guideImagesArr.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    NewFeatureCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:reuseIdentifier forIndexPath:indexPath];
    cell.startImage = [UIImage imageNamed:self.guideImagesArr[indexPath.row]];
    cell.moviePath = self.guideMoviePathArr[indexPath.row];
    self.movieFinished = NO;
    return cell;
}

#pragma mark <UICollectionViewDelegate>

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.row == self.guideImagesArr.count - 1 && self.isMovieFinished) {
        if (self.lastOnePlayFinished) {
            self.lastOnePlayFinished();
        }  
    }
}

#pragma mark <UIScrollViewDelegate>

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    uint page = (scrollView.contentOffset.x / scrollView.bounds.size.width) + 0.5;
    self.pageControl.currentPage = page;
}

@end

File: FleaMarketGuideSwift/FleaMarketGuideSwift/Classes/NewFeature/Controller/NewFeatureViewController.swift
//
//  NewFeatureViewController.swift
//  FleaMarketGuideSwift
//
//  Created by ALin on 16/6/12.
//  Copyright © 2016年 ALin. All rights reserved.
//

import UIKit

let MS_Width = UIScreen.mainScreen().bounds.size.width
let MS_Height = UIScreen.mainScreen().bounds.size.height
let Key_Window = UIApplication.sharedApplication().keyWindow

private let reuseIdentifier = "NewFeatureCell"

class NewFeatureViewController: UICollectionViewController {
    
    var guideImages: [UIImage]
    
    var guideMoviePaths: [String]
    
    var lastOnePlayFinished: ()->()
    
    // MARK: - life circle
    
    /**
     构造方法
     
     - parameter images:       封面图片
     - parameter moviePaths:   视频地址
     - parameter playFinished: 最后一个视频播放完毕
     */
    init(images:[UIImage], moviePaths:[String], playFinished: ()->()) {
        guideImages = images
        guideMoviePaths = moviePaths
        lastOnePlayFinished = playFinished
        
        super.init(collectionViewLayout: ALinFlowLayout())
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        self.collectionView!.registerClass(NewFeatureCell.self, forCellWithReuseIdentifier: reuseIdentifier)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(NewFeatureViewController.finishedPlay), name: PlayFinishedNotify, object: nil)
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        pageControl.numberOfPages = guideMoviePaths.count ?? 0
        
    }
    
    deinit
    {
        pageControl.removeFromSuperview()
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    // MARK: - private method
    
    private var isMovieFinished = false
    func finishedPlay() {
        isMovieFinished = true
    }

    // MARK: UICollectionViewDataSource
    
    override func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
       return guideMoviePaths.count ?? 0
    }

    override func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier(reuseIdentifier, forIndexPath: indexPath) as! NewFeatureCell
        
        cell.coverImage = guideImages[indexPath.item]
        
        cell.moviePath = guideMoviePaths[indexPath.item]
        
        isMovieFinished = false
        return cell
    }
    
    // MARK: UICollectionViewDelegate
    override func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
        if indexPath.item ==  guideMoviePaths.count - 1 && isMovieFinished{
            lastOnePlayFinished()
        }
    }
    
    // MARK: UIScrollViewDelegate
    override func scrollViewDidScroll(scrollView: UIScrollView) {
        let page = Int(scrollView.contentOffset.x / collectionView!.bounds.width + 0.5)
        pageControl.currentPage = page
    }

    // MARK : - 懒加载
    private lazy var pageControl : UIPageControl =
    {
        let width : CGFloat = 120.0
        let height : CGFloat = 30.0
        let x : CGFloat = (MS_Width - width) * 0.5;
        let y : CGFloat =  MS_Height - 30 - 20;
        let pageControl = UIPageControl(frame: CGRect(x: x, y: y, width: width, height: height))
            pageControl.pageIndicatorTintColor = UIColor.lightGrayColor()
        pageControl.currentPageIndicatorTintColor = UIColor.redColor()
        Key_Window?.addSubview(pageControl)
        return pageControl;
    }()
    
    
}
